tasks/src/tasks/fetch.rs

```diff
@@ -0,0 +1,137 @@
+use std::{collections::HashMap, num::NonZeroU32};
+
+use imap_types::{
+    command::CommandBody,
+    core::Vec1,
+    fetch::{MacroOrMessageDataItemNames, MessageDataItem},
+    response::{Data, StatusBody, StatusKind},
+    sequence::{SeqOrUid, SequenceSet},
+};
+use tracing::warn;
+
+use super::TaskError;
+use crate::Task;
+
+/// Fetch message data items matching the given sequence set and the
+/// given item names (or macro).
+#[derive(Clone, Debug)]
+pub struct FetchTask {
+    sequence_set: SequenceSet,
+    macro_or_item_names: MacroOrMessageDataItemNames<'static>,
+    uid: bool,
+    output: HashMap<NonZeroU32, Vec1<MessageDataItem<'static>>>,
+}
+
+impl Task for FetchTask {
+    type Output = Result<HashMap<NonZeroU32, Vec1<MessageDataItem<'static>>>, TaskError>;
+
+    fn command_body(&self) -> CommandBody<'static> {
+        CommandBody::Fetch {
+            sequence_set: self.sequence_set.clone(),
+            macro_or_item_names: self.macro_or_item_names.clone(),
+            uid: self.uid,
+        }
+    }
+
+    fn process_data(&mut self, data: Data<'static>) -> Option<Data<'static>> {
+        if let Data::Fetch { items, seq } = data {
```

Comment:
Here I have a proposition. It may be a bit too opinated, yet I found it relevant to share.

If you use `uid` for this task, the output still returns a hash map of sequence. It could be nice to return either UID or seq depending on `uid: bool`, but it implies to modify the item names provided by the user (pushing the `Uid` item name). Here how it looks like:

```rust
// First expand items, then push Uid if needed
let mut fetch_items = match fetch_items {
    MacroOrMessageDataItemNames::Macro(m) => m.expand().into_static(),
    MacroOrMessageDataItemNames::MessageDataItemNames(items) => items,
};

if uid {
    fetch_items.push(MessageDataItemName::Uid);
}

// In process_data, we can now extract the Uid from the items
fn process_data(&mut self, data: Data<'static>) -> Option<Data<'static>> {
    if let Data::Fetch { items, seq } = data {
        if self.uid {
            let uid = items.as_ref().iter().find_map(|item| {
                if let MessageDataItem::Uid(uid) = item {
                    Some(*uid)
                } else {
                    None
                }
            });

	    // We could even unwrap since we know that MessageDataItemName::Uid
	    // have been pushed to the message item names
	    
            if let Some(uid) = uid {
                self.output.insert(uid, items);
            } else {
                warn!("cannot get UID of message {seq}, skipping fetch data");
            }
        } else {
            self.output.insert(seq, items);
        }

        None
    } else {
        Some(data)
    }
}
```

Reply:
I like the idea but it has some downsides :-(

The data we get for sure is the `seq`. Everything else might not be present -- even if we ask for it (This is also why we should not use `unwrap`: A broken (or even malicious) server must not be able to crash us.) Thus, what's left is to not include the UID in the hash map (and maybe notify the user) -- exactly what you propose.

The downside is that now we alter the session. Instead of saying "UID missing", we say "fetch not sent". Not sure how important that is, but it feels "surprising". Also, we remove the `seq` which is, I feel, probably better explicitly removed at the user site.

Maybe these downsides can be resolved with a different design. But I feel that the fact that UID may not be present prevents a clean API :-(

Reply:
Makes total sense. What is bothering me the most is the lack of consistency among tasks using either seq or uid. Let's take the `SEARCH`, `SORT`, and `THREAD` tasks: the `uid: bool` parameter affects *input* (defines if sequence set is made of seq or uid) and *output* (returns a vector of seq or uid). In the case of `FETCH` (and `STORE` as well), `uid: bool` only affects *input*: the *output* uses seq no matter what. But after all, it is the way the protocol is defined, so let stick to it.

> Everything else might not be present -- even if we explicitly ask for it.

Sure, but you have the exact same issue if you request manually the uid. A user who `FETCH` using uid implicitly set should be as surprised as the one who `FETCH` using uid explicitly set. So I ask you this question: what is the best way we could handle this case of missing data explicitly (or implitictly) set? Let's say only one `FETCH` does not have uid, should we send all other fetches and notify the missing one, or fail? By experience, I would go for the first option, at least in does not interrupt the user.

Another idea that just came to by mind is to return a struct (or enum) that contains both correct and potentially incorrect output. It is a bit more verbose version of the log warn notification, but at least user has total control of the situation. A good package is [`These`](https://github.com/FintanH/these), the "inclusive or" version of `Result`.

```rust
type Output = Result<These<Vec<MessageDataItem>, Vec<MessageDataItem>>, TaskError>
```

What a mess :smiling_face_with_tear:

Reply:
Yeah, it's a mess. Let's discuss this at a later time. `FetchFirstTask` is not too repetitve. Let's get this in amd see how it goes.